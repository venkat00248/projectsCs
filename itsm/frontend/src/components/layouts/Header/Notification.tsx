import React from "react";
import "./Notification.scss";
import { useDispatch } from "react-redux";
import io from 'socket.io-client';
import { ActionTypes } from "../../../redux/constants/action-types";
import { useTranslation } from "react-i18next";
import { useSelector } from 'react-redux';
import { getColorForLetter } from "../../Utils/GetColorNotification";
import { config } from "../../../config/config";

const socket = io(config.ENV.PROXY_API_URL);
const Notification = ({notifications}:any) => {
    const dispatch = useDispatch()
    const { t } = useTranslation();
    const toogleNotification = useSelector((state:any) => state.toogleNotification);
    const handleRead = (id?:any) => {
      socket.emit('markNotification', { type: 'single', notificationId: id });
    }
    const handelNotiDetails = (ticketID:any, id?:any, status?:any) => {
      dispatch({ type: ActionTypes.SET_TOOGLE_NOTIFICATION, payload: {flag: false}});
      sessionStorage.setItem("ticketId", ticketID);
      if(status == false) {
        socket.emit('markNotification', { type: 'single', notificationId: id });
      }
    }

    return(
        <div>
          <section className="section-10 pt-0">
              <div className="container">
                <div className="notification-ui_dd-content"  data-testid = "notificationContainer">
                  {
                    notifications?.map((item:any, index:any) => (
                      <div className={`notification-list ${item?.read? '':'markas-not-read'}`} key={index} data-testid = "notificationItem">
                      <div className="notification-list_content d-flex" style={{cursor: "pointer"}} onClick={() => handelNotiDetails(item?.ticketID, item?.id, item?.read)} >
                          <div className="flex-shrink-0 me-3">
                                <div className="avatar-xs">
                                  <span className={`avatar-title font-size-16 ${getColorForLetter(item?.username)}`}>
                                    <i className="fst-normal">{item?.username.charAt(0)}</i>
                                  </span>
                                </div>
                          </div>
                          <div className="flex-grow-1">
                              <div className="assigned text-secondary">
                              <i className="mdi mdi-ticket-confirmation-outline verticalicon"></i>
                              <span className="mx-3 ">Ticket Assigned</span>
                              <span className="text-success border px-2 rounded dateandtime">{item?.ticketActivity}</span>
                              </div>
  
                              <div className="username">
                                <p className="fw-bold text-dark"> 
                                    <span>{item?.ticketID}</span> - <span>(</span><span>{item?.ticketStatus}</span><span>)</span> assigned to
                                </p>
                                <span className="text-muted">{item?.username}</span>
                                <p className="fw-bold text-muted">Cloud4c</p>
                              </div>
  
                            </div>
                      </div>
                      <div className="d-flex align-items-center mbl-mark">
                        <span className={`markRead ${item?.read? "read-Color":"text-primary"}`} onClick={() => handleRead(item?.id)} data-testid= "markReadButton">
                              {t(item?.action)}
                        </span> 
                      </div>
                    </div>
                    ))
                  }
                </div>
              </div>
          </section>
        </div>
    )
} 
export default Notification;