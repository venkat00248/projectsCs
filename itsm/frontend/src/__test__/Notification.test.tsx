import { fireEvent, screen, render as rtlrender } from "@testing-library/react";
import { Provider } from 'react-redux';
import Notification from '../components/layouts/Header/Notification';
import store from '../redux/Store';
import renderer from 'react-test-renderer'

const renderWithProvider = (component:any) =>
  rtlrender(<Provider store={store}>{component}</Provider>);

describe('Notification component', () => {
  const notifications = [
    {
      id: 1,
      username: "Karteek Vemula",
      ticketID: "TSK101233792",
      ticketActivity: "24th May 2023 5:17 PM",
      ticketStatus: "Open",
      dateTime: "Just now",
      action: "Mark as read",
      read: false
    }
  ];
  
  it('render notificationContainer ID in the document', () => {
    const wrapper = renderWithProvider(<Notification notifications= {notifications}/>);
    const ntfcContainer = wrapper.getByTestId("notificationContainer");
    expect(ntfcContainer).toBeInTheDocument();
  });
  it('Verify that each notification item is rendered correctly', () => {
    const wrapper = renderWithProvider(<Notification notifications= {notifications}/>);
    const notificationItems = wrapper.getAllByTestId('notificationItem');
    expect(notificationItems).toHaveLength(notifications.length);
  })
  it('Verify the correct class is applied based on the read status and field', () => {
    const wrapper = renderWithProvider(<Notification notifications= {notifications}/>);
    const notificationItems = wrapper.getAllByTestId('notificationItem');
    notifications.forEach((notification, index) => {
      const notificationItem = notificationItems[index];

      // Verify correct class is applied or not //
      const expectedClass = notification.read ? 'notification-list' : 'notification-list markas-not-read';
      expect(notificationItem).toHaveClass(expectedClass);

      // Verify username in the document //
      const usernameElement = screen.getByText(notification.username);
      expect(usernameElement).toBeInTheDocument();
      // Verify ticketID in the document //
      const ticketIdElement = screen.getByText(notification.ticketID);
      expect(ticketIdElement).toBeInTheDocument();
      // Verify ticketActivity in the document //
      const ticketActivityElement = screen.getByText(notification.ticketActivity);
      expect(ticketActivityElement).toBeInTheDocument();
      // Verify ticketStatus in the document //
      const ticketStatusElement = screen.getByText(notification.ticketStatus);
      expect(ticketStatusElement).toBeInTheDocument();
    });
  });
  it('should handle read action', () => {
    const wrapper = renderWithProvider(<Notification notifications= {notifications}/>);
    const markReadAction = wrapper.getByTestId('markReadButton');
    expect(markReadAction).toBeInTheDocument();
  });
  it('should call handleRead function on button click', () => {
    const wrapper = renderWithProvider(<Notification notifications={notifications} />);
    const markReadButton = wrapper.getByTestId('markReadButton');
    expect(fireEvent.click(markReadButton)).toBeTruthy()
  });
  it('should render with given state from Redux store', () => {
    const wrapper = renderer.create(
        <Provider store={store}>
            <Notification notifications={notifications} />
        </Provider>
    )
    expect(wrapper.toJSON()).toMatchSnapshot();
});
});