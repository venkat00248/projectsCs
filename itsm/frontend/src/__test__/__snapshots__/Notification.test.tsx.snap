// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component should render with given state from Redux store 1`] = `
<div>
  <section
    className="section-10 pt-0"
  >
    <div
      className="container"
    >
      <div
        className="notification-ui_dd-content"
        data-testid="notificationContainer"
      >
        <div
          className="notification-list markas-not-read"
          data-testid="notificationItem"
        >
          <div
            className="notification-list_content d-flex"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <div
              className="flex-shrink-0 me-3"
            >
              <div
                className="avatar-xs"
              >
                <span
                  className="avatar-title font-size-16 bg-K"
                >
                  <i
                    className="fst-normal"
                  >
                    K
                  </i>
                </span>
              </div>
            </div>
            <div
              className="flex-grow-1"
            >
              <div
                className="assigned text-secondary"
              >
                <i
                  className="mdi mdi-ticket-confirmation-outline verticalicon"
                />
                <span
                  className="mx-3 "
                >
                  Ticket Assigned
                </span>
                <span
                  className="text-success border px-2 rounded dateandtime"
                >
                  24th May 2023 5:17 PM
                </span>
              </div>
              <div
                className="username"
              >
                <p
                  className="fw-bold text-dark"
                >
                  <span>
                    TSK101233792
                  </span>
                   - 
                  <span>
                    (
                  </span>
                  <span>
                    Open
                  </span>
                  <span>
                    )
                  </span>
                   assigned to
                </p>
                <span
                  className="text-muted"
                >
                  Karteek Vemula
                </span>
                <p
                  className="fw-bold text-muted"
                >
                  Cloud4c
                </p>
              </div>
            </div>
          </div>
          <div
            className="d-flex align-items-center mbl-mark"
          >
            <span
              className="markRead text-primary"
              data-testid="markReadButton"
              onClick={[Function]}
            >
              Mark as read
            </span>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
